Cubical/Foundations/HAEquiv.agda

```diff
@@ -0,0 +1,120 @@
+
+open import Agda.Primitive
+open import Cubical.Core.Prelude
+open import Cubical.Core.Glue
+open import Cubical.Foundations.Isomorphism
+open import Cubical.Foundations.Equiv
+open import Cubical.Foundations.Univalence
+
+record isHAEquiv {ℓ ℓ'} {A : Set ℓ} {B : Set ℓ'} (f : A → B) : Set (ℓ ⊔ ℓ') where
+  field
+    g : B → A
+    sec : ∀ a → g (f a) ≡ a
+    ret : ∀ b → f (g b) ≡ b
+    com : ∀ a → cong f (sec a) ≡ ret (f a)
+
+HAEquiv : ∀ {ℓ ℓ'} (A : Set ℓ) (B : Set ℓ') → Set (ℓ ⊔ ℓ')
+HAEquiv A B = Σ (A → B) λ f → isHAEquiv f
+
+private
+  variable
+    ℓ ℓ' : Level
+    A : Set ℓ
+    B : Set ℓ'
+
+homotopyNatural' : {f g : A → B} (H : ∀ a → f a ≡ g a) {x y : A} (p : x ≡ y) → (H x) ∙ (cong g p) ≡ (cong f p) □ (H y)
+homotopyNatural' {f = f} {g = g} H {x} {y} p i j =
+  hcomp (λ k → \ { (i = i0) → compPath-filler (H x) (cong g p) k j
+                 ; (i = i1) → compPath'-filler (cong f p) (H y) k j
+                 ; (j = i0) → cong f p (i ∧ (~ k))
+                 ; (j = i1) → cong g p (i ∨ k) })
+        (H (p i) j)
+
+homotopyNatural : {f g : A → B} (H : ∀ a → f a ≡ g a) {x y : A} (p : x ≡ y) → (H x) ∙ (cong g p) ≡ (cong f p) ∙ (H y)
+homotopyNatural H p = (homotopyNatural' H p) ∙ (□≡∙ _ _)
+
+Hfa≡fHa : ∀ {ℓ} {A : Set ℓ} (f : A → A) (H : ∀ a → f a ≡ a) → ∀ a → H (f a) ≡ cong f (H a)
+Hfa≡fHa {A = A} f H a = H (f a) ≡⟨ sym (∙-rUnit (H (f a))) ⟩
+                        (H (f a)) ∙ refl ≡⟨ cong (_∙_ (H (f a))) (sym (∙-rInv (H a))) ⟩
+                        (H (f a)) ∙ ((H a) ∙ (sym (H a))) ≡⟨ sym (∙-assoc _ _ _ )⟩
+                        ((H (f a)) ∙ (H a)) ∙ (sym (H a)) ≡⟨ cong (λ x →  x ∙ (sym (H a))) (homotopyNatural H (H a)) ⟩
+                        ( (cong f (H a)) ∙ (H a)) ∙ (sym (H a)) ≡⟨ ∙-assoc _ _ _ ⟩
+                        (cong f (H a)) ∙ ((H a) ∙ (sym (H a))) ≡⟨ cong (_∙_ (cong f (H a))) (∙-rInv _) ⟩
+                        (cong f (H a)) ∙ refl ≡⟨ ∙-rUnit _ ⟩
+                        cong f (H a) ∎
+
+iso→HAEquiv : Iso A B → HAEquiv A B
+iso→HAEquiv {A = A} {B = B} (iso f g ε η) = f , (record { g = g ; sec = η ; ret = ret ; com = com })
+  where
+    sides : ∀ b i j → Partial (~ i ∨ i) B
+    sides b i j = \ { (i = i0) → ε (f (g b)) j
+                   ; (i = i1) → ε b j }
+
+    bot : ∀ b i → B
+    bot b i = cong f (η (g b)) i
+
+    ret : (b : B) → f (g b) ≡ b
+    ret b i = hcomp (sides b i) (bot b i)
+
+    com : (a : A) → cong f (η a) ≡ ret (f a)
+    com a i j = hcomp (λ k → \ { (i = i0) → ε (f (η a j)) k
+                                ; (i = i1) → hfill (sides (f a) j) (inc (bot (f a) j)) k
+                                ; (j = i0) → ε (f (g (f a))) k
+                                ; (j = i1) → ε (f a) k})
+                       ( cong (cong f) (sym (Hfa≡fHa (λ x → g (f x)) η a)) i j)
+
+equiv→HAEquiv : A ≃ B → HAEquiv A B
+equiv→HAEquiv e = iso→HAEquiv (equivToIso e)
+
+congEquiv : ∀ {ℓ ℓ'} {A : Set ℓ} {B : Set ℓ'} {x y : A} (e : A ≃ B) → (x ≡ y) ≃ (e .fst x ≡ e .fst y)
```

Comment:
This is proved somewhere else as discussed, but that proof was not as general? (A and B had to be in the same universe)

Maybe this proof is the one that should be used everywhere?